runtime-modules/storage/src/lib.rs

```diff
@@ -1,20 +1,2373 @@
+            Error::<T>::VoucherMaxObjectNumberLimitExceeded
+        );
+
+        if let Some(operator_id) = invited_worker {
+            Self::ensure_storage_provider_operator_exists(operator_id)?;
+        }
+
+        Ok(())
+    }
+
+    // Update total objects size and number for all storage buckets assigned to a bag.
+    fn change_storage_bucket_vouchers_for_bag(
+        bag_id: &BagId<T>,
+        voucher_update: &VoucherUpdate,
+        voucher_operation: OperationType,
+    ) {
+        let bucket_ids = BagManager::<T>::get_storage_bucket_ids(bag_id);
+
+        Self::change_storage_buckets_vouchers(&bucket_ids, voucher_update, voucher_operation);
+    }
+
+    // Update total objects size and number for provided storage buckets.
+    fn change_storage_buckets_vouchers(
+        bucket_ids: &BTreeSet<T::StorageBucketId>,
+        voucher_update: &VoucherUpdate,
+        voucher_operation: OperationType,
+    ) {
+        for bucket_id in bucket_ids.iter() {
+            <StorageBucketById<T>>::mutate(bucket_id, |bucket| {
+                bucket.voucher =
+                    voucher_update.get_updated_voucher(&bucket.voucher, voucher_operation);
+
+                Self::deposit_event(RawEvent::VoucherChanged(*bucket_id, bucket.voucher.clone()));
+            });
+        }
+    }
+
+    // Validates `delete_data_objects` parameters.
+    // Returns voucher update for an affected bag.
+    fn validate_delete_data_objects_params(
+        bag_id: &BagId<T>,
+        data_object_ids: &BTreeSet<T::DataObjectId>,
+    ) -> Result<BagChangeInfo<BalanceOf<T>>, DispatchError> {
+        ensure!(
+            !data_object_ids.is_empty(),
+            Error::<T>::DataObjectIdParamsAreEmpty
+        );
+
+        BagManager::<T>::ensure_bag_exists(bag_id)?;
+
+        let mut bag_change = BagChangeInfo::default();
+
+        for data_object_id in data_object_ids.iter() {
+            let data_object =
+                BagManager::<T>::ensure_data_object_existence(bag_id, data_object_id)?;
+
+            bag_change.add_object(data_object.size, data_object.deletion_prize);
+        }
+
+        ensure!(
+            <StorageTreasury<T>>::usable_balance() >= bag_change.total_deletion_prize,
+            Error::<T>::InsufficientTreasuryBalance
+        );
+
+        Ok(bag_change)
+    }
+
+    // Validates upload parameters and conditions (like global uploading block).
+    // Returns voucher update parameters for the storage buckets.
+    fn validate_upload_data_objects_parameters(
+        params: &UploadParameters<T>,
+    ) -> Result<BagChangeInfo<BalanceOf<T>>, DispatchError> {
+        // Check global uploading block.
+        ensure!(!Self::uploading_blocked(), Error::<T>::UploadingBlocked);
+
+        // Check object creation list validity.
+        ensure!(
+            !params.object_creation_list.is_empty(),
+            Error::<T>::NoObjectsOnUpload
+        );
+
+        BagManager::<T>::ensure_bag_exists(&params.bag_id)?;
```

Comment:
Ok. BagManager has a state now (bag_id), and I combined its construction with an existence check.

Reply:
Does this really fix the problem? If you return a `BagManager`, you have lost the static guarantee that the bag exists, which was the main issue to begin with. Seems to me that if the way a manager is used is that its always kept around after a bag has been shown to exist, then you need to keep a copy in the manager, however, you are mostly not using these manager instances in a statefule way, so I am not sure that really makes sense. Perhaps the semantics of what this manager represents i a bit unclear, let me know what you think, and just make the decision you think is best.

Reply:
The `BagManager` is an abstraction over the static and dynamic bags storage and types. 

It handles different structs for `DynamicBag` and `StaticBag` and different storage maps for static and dynamic bags. Currently, there are only two types of storage for bags. One combined storage map for the static bags (council and working groups) and one for all dynamic bags(Member and Channel). But we could fall back to the initial storage design with one storage map per bag type without changing the code that uses the BagManager.

For example, instead of 

`BagManager::<T>::ensure_bag_exists(&bag_id)?;`

we could write

```
match bag_id{
    Static(static_bag_id) => ensure_static_bag_exists(static_bag_id)
    Dynamic(dynamic_bag_id) => ensure_dynamic_bag_exists(dynamic_bag_id)
}?;
```

The code for getting the bag, manipulating with its content and saving back will be the same for all storages. So, I absracted it with the `Bagmanager`.

However, this all goes away if we drop the `deletion prize` field in the dynamic bag struct (your another comment). Because, it makes the `StaticBag` struct identical with the `DynamicBag` struct. And I can even create a single storage map `BagId => Bag` and remove a huge part of `BagManager` or drop it completely.

Reply:
1. I understand, that is indeed practical, I think my point in that particular case would be to return some sum type across the two bag types, so that the ensure abstraction is similar to how we typically do it. But I dont know if that causes other problems later.
2. I suggested a way to approach having just a single bag type here https://github.com/Joystream/joystream/pull/2407#discussion_r676404403, if that sounds advisable, then lets do that. However, let us not put static bags into the dynamic map, as that makes the representations much less safe, as static bags can be deleted and so on. Let just use the same approach as now, but with only one bag type.